staging/src/k8s.io/apiserver/pkg/storage/etcd/api_object_version_obfuscator.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package etcd
+
+import (
+	"encoding/binary"
+	"hash/fnv"
+)
+
+// FeistelObfuscator implements Obfuscator.
+type FeistelObfuscator struct {
+	KeyScheduleFunc func(key string) []uint32
+	RoundFunc       func(block uint32, roundKey uint32) uint32
+}
+
+// NewFeistelObfuscator instantiates a FeistelObfuscator object.
+func NewFeistelObfuscator() FeistelObfuscator {
+	return FeistelObfuscator{
+		KeyScheduleFunc: HashKeySchedule,
+		RoundFunc:       ProductRoundFunc,
+	}
+}
+
+// HashKeySchedule takes the 128 hash of a string and breaks it
+// into 4 32-bit keys to be used as a key schedule.
+func HashKeySchedule(key string) []uint32 {
+	hash := fnv.New128a()
+	hash.Write([]byte(key))
+	hashBytes := hash.Sum(make([]byte, 0))
+	roundKeys := make([]uint32, 0)
+	for i := 0; i < len(hashBytes); i += 4 {
+		roundKey := binary.BigEndian.Uint32(hashBytes[i : i+4])
+		roundKeys = append(roundKeys, roundKey)
+	}
+	return roundKeys
+}
+
+// ProductRoundFunc is a round function that just multiplies the values
+// of the block and the roundKey.
+func ProductRoundFunc(block uint32, roundKey uint32) uint32 {
+	return block * roundKey
+}
+
+// Decode implements Obfuscator.
+func (f FeistelObfuscator) Decode(key string, clientResourceVersion uint64) uint64 {
+	left, right := splitUint(clientResourceVersion)
+	keySchedule := f.KeyScheduleFunc(key)
```

Comment:
When we decide to implement keys as the collection name, then precomputing wouldn't make sense anymore. If we tried to cache it in a map, the get operation from the map would start by computing a hash of the string, so I figured this would be just as fast as that.

Should it precompute it for now though, and change to something like this in a future PR?

Or maybe each collection could somehow store the precomputed key schedule.